java-checks/src/main/java/org/sonar/java/checks/FileCreateTempFileCheck.java

```diff
@@ -0,0 +1,149 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.MethodInvocationMatcher;
+import org.sonar.plugins.java.api.JavaFileScanner;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.ParenthesizedTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+
+@Rule(
+  key = "S2976",
+  name = "\"File.createTempFile\" should not be used to create a directory",
+  tags = {"owasp-a9", "security"},
+  priority = Priority.CRITICAL)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.API_ABUSE)
+@SqaleConstantRemediation("5min")
+public class FileCreateTempFileCheck extends BaseTreeVisitor implements JavaFileScanner {
+
+  private enum State {
+    CREATE_TMP_FILE,
+    DELETE,
+    MKDIR
+  }
+
+  private static final String JAVA_IO_FILE = "java.io.File";
+  private static final MethodInvocationMatcher FILE_CREATE_TEMP_FILE = MethodInvocationMatcher.create()
+    .typeDefinition(JAVA_IO_FILE).name("createTempFile").withNoParameterConstraint();
+  private static final MethodInvocationMatcher FILE_DELETE = MethodInvocationMatcher.create()
+    .typeDefinition(JAVA_IO_FILE).name("delete");
+  private static final MethodInvocationMatcher FILE_MKDIR = MethodInvocationMatcher.create()
+    .typeDefinition(JAVA_IO_FILE).name("mkdir");
+
+  private final Deque<Map<Symbol, State>> symbolStack = new LinkedList<>();
+  private JavaFileScannerContext context;
+
+  @Override
+  public void scanFile(final JavaFileScannerContext context) {
+    this.context = context;
+    scan(context.getTree());
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    symbolStack.push(new HashMap<Symbol, State>());
```

Comment:
I am not sure of the use of a stack here. What's the point ? to delimitate the calls within a method ? Then there should be a unit test showing its utility with an inner class.